File: app/src/main/java/com/apps/adrcotfas/goodtime/Util/StringUtils.java
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.apps.adrcotfas.goodtime.Util;

import org.joda.time.DateTime;
import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.Locale;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;

public class StringUtils {

    private static final DateTimeFormatter monthFormatter = DateTimeFormat.forPattern("EEE', 'MMM d', ' yyyy");
    private static final DateTimeFormatter timeFormatter = DateTimeFormat.forPattern("HH:mm");
    private static final DateTimeFormatter backUpFormatter = DateTimeFormat.forPattern("yyyy-MM-dd-HH:mm");
    private static final DateTimeFormatter dayOfWeekFormatter = DateTimeFormat.forPattern("E");
    private static final DateTimeFormatter hourOfDayFormatter = DateTimeFormat.forPattern("hh");
    private static final DateTimeFormatter meridianFormatter = DateTimeFormat.forPattern("a");

    private static final NavigableMap<Long, String> suffixes = new TreeMap<>();
    static {
        suffixes.put(1_000L, "k");
        suffixes.put(1_000_000L, "M");
        suffixes.put(1_000_000_000L, "G");
        suffixes.put(1_000_000_000_000L, "T");
        suffixes.put(1_000_000_000_000_000L, "P");
        suffixes.put(1_000_000_000_000_000_000L, "E");
    }

    /**
     * Shortens a value in minutes to a easier to read format of maximum 4 characters
     * Do not use this for anything else than the History chart before removing the extra left padding hack
     * @param value the value in minutes to be formatted
     * @return the formatted value
     */
    public static String formatLong(long value) {
        //Long.MIN_VALUE == -Long.MIN_VALUE so we need an adjustment here

        if (value == Long.MIN_VALUE) return formatLong(Long.MIN_VALUE + 1);
        if (value < 0) return "-" + formatLong(-value);
        if (value < 1000) return Long.toString(value); //deal with easy case

        Map.Entry<Long, String> e = suffixes.floorEntry(value);
        Long divideBy = e.getKey();
        String suffix = e.getValue();

        long truncated = value / (divideBy / 10); //the number part of the output times 10
        boolean hasDecimal = truncated < 100 && (truncated / 10d) != (truncated / 10);

        return hasDecimal ? (truncated / 10d) + suffix : (truncated / 10) + suffix;
    }

    public static String formatMinutes(long minutes) {

        final long days   = minutes / 1440;
        final long hours  = minutes / 60 % 24;
        final long remMin = minutes % 60;

        String result;
        if (minutes != 0) {
            result = (days != 0 ? "" + Long.toString(days) + "d" + "\n" : "")
                    + (hours != 0 ? "" + Long.toString(hours) + "h" : "")
                    + (remMin != 0 ? " " + Long.toString(remMin) + "m" : "");
        } else {
            result = "0m";
        }
        return result;
    }

    public static String formatDate(long millis) {
        return monthFormatter.print(millis);
    }

    public static String formatTime(long millis) {
        return timeFormatter.print(millis);
    }

    public static String formatDateAndTime(long millis) {
        return backUpFormatter.print(millis);
    }

    public static String toPercentage(float value) {
        return Math.round(100 * value) + "%";
    }

    public static String toDayOfWeek(int value) {
        return dayOfWeekFormatter.withLocale(Locale.getDefault()).print(new DateTime().withDayOfWeek(value));
    }

    public static String toHourOfDay(int value) {
        DateTime date = new DateTime().withHourOfDay(value);
        return hourOfDayFormatter.print(date) + "\n" + meridianFormatter.print(date);
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/util/StringUtils.kt
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.util

import java.time.*
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.time.temporal.TemporalAdjusters
import java.util.*
import kotlin.math.roundToInt

object StringUtils {

    private val backUpFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm")
    private val dayOfWeekFormatter = DateTimeFormatter.ofPattern("E")
    private val suffixes: NavigableMap<Long, String> = TreeMap()

    /**
     * Shortens a value in minutes to a easier to read format of maximum 4 characters
     * Do not use this for anything else than the History chart before removing the extra left padding hack
     * @param value the value in minutes to be formatted
     * @return the formatted value
     */
    fun formatLong(value: Long): String {
        //Long.MIN_VALUE == -Long.MIN_VALUE so we need an adjustment here
        if (value == Long.MIN_VALUE) return formatLong(Long.MIN_VALUE + 1)
        if (value < 0) return "-" + formatLong(-value)
        if (value < 1000) return value.toString() //deal with easy case
        val e = suffixes.floorEntry(value)!!
        val divideBy = e.key
        val suffix = e.value
        val truncated = value / (divideBy / 10) //the number part of the output times 10
        val hasDecimal = truncated < 100 && truncated / 10.0 != (truncated / 10).toDouble()
        return if (hasDecimal) (truncated / 10.0).toString() + suffix else (truncated / 10).toString() + suffix
    }

    fun formatDateAndTime(millis: Long): String {
        return millis.toLocalDateTime().format(backUpFormatter)
    }

    @JvmStatic
    fun formatTimeForStatistics(millis: Long?): String {
        val date = LocalDateTime.ofInstant(
            Instant.ofEpochMilli(
                millis!!
            ), ZoneId.systemDefault()
        )
        return date.format(DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT))
    }

    @JvmStatic
    fun formatDateForStatistics(millis: Long?): String {
        val date = LocalDateTime.ofInstant(
            Instant.ofEpochMilli(
                millis!!
            ), ZoneId.systemDefault()
        )
        return "${date.format(DateTimeFormatter.ofPattern("EEEE"))} ${date.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT))}"
    }

    fun toPercentage(value: Float): String {
        return (100 * value).roundToInt().toString() + "%"
    }

    fun toDayOfWeek(value: Int): String {
        return LocalDateTime.now().with(TemporalAdjusters.next(DayOfWeek.of(value))).format(dayOfWeekFormatter)
    }

    fun toHourOfDay(hour: Int, is24HourFormat: Boolean): String {
        val time = LocalTime.now().withHour(hour)
        return time.format(
            DateTimeFormatter.ofPattern(
                if (is24HourFormat) "HH"
                else "hh a"
            )
        ).replace(" ", "\n")
    }

    init {
        suffixes[1000L] = "k"
        suffixes[1000000L] = "M"
        suffixes[1000000000L] = "G"
        suffixes[1000000000000L] = "T"
        suffixes[1000000000000000L] = "P"
        suffixes[1000000000000000000L] = "E"
    }

    fun formatMinutes(minutes: Long): String {
        val days = minutes / 1440
        val hours = minutes / 60 % 24
        val remMin = minutes % 60
        val result: String = if (minutes != 0L) {
            ((if (days != 0L) "${days}d\n" else "")
                    + (if (hours != 0L) "" + hours.toString() + "h" else "")
                    + if (remMin != 0L) " " + remMin.toString() + "m" else "")
        } else {
            "0m"
        }
        return result
    }
}